packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -0,0 +1,149 @@
+import { Construct } from './construct';
+import { Token } from './tokens';
+
+/**
+ * ITaggable indicates a entity manages tags via the `tags` property
+ */
+export interface ITaggable {
+    tags: TagManager,
+}
+
+/**
+ * A single tag property values
+ */
+interface ITagProps {
+    value: string | Token;
+    propagate?: boolean;
+}
+
+/**
+ * A an object of tags
+ */
+export interface ITags {
+    [key: string]: ITagProps;
+}
+
+/**
+ * A single tag
+ */
+export interface ITag {
+    key: string;
+    value: string | Token;
+}
+
+/**
+ * TagManager facilitates a common implementation of tagging for Constructs.
+ *
+ * Each construct that wants to support tags should implement the `ITaggable`
+ * interface and properly pass tags to the `Resources` (Cloudformation) elements
+ * the `Construct` creates using `toToken()` for lazy evaluations
+ */
+export class TagManager {
+
+    /**
+     * Checks if the object implements the `ITaggable` interface
+     */
+    public static isTaggable(taggable: ITaggable | any): taggable is ITaggable {
+        return ((taggable as ITaggable).tags !== undefined);
+    }
+
+    private readonly _tags: ITags = {};
+
+    private readonly parentTagManager: TagManager | undefined;
+
+    constructor(partner: Construct, initialTags: ITags = {}) {
+        if (TagManager.isTaggable(partner.parent)) {
+            this.parentTagManager = partner.parent.tags;
```

Comment:
What happens if the parent is not taggable, but the grand-parent is (etc). Should probably recurse until you find someone who is taggable...

Reply:
Yep this was by design, but you are right it should not stop at the first non-taggable parent. Fixed by changing the logic with `ancestors` and test cases added.